c2rust-analyze/src/known_fn.rs

```diff
@@ -0,0 +1,362 @@
+use crate::context::PermissionSet;
+
+#[cfg(test)]
+macro_rules! const_slice {
+    ($ty:ty, []) => {{
+        &[]
+    }};
+    ($ty:ty, $array:expr) => {{
+        const ARRAY: [$ty; $array.len()] = $array;
+        &ARRAY
+    }};
+}
+
+#[cfg(test)]
+macro_rules! perms_annotation {
+    ([$($($perm:ident)|*),*]) => {{
+        [$(PermissionSet::union_all([$(PermissionSet::$perm,)*]),)*]
+    }};
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub struct KnownFnTy {
+    name: &'static str,
+    ty: &'static str,
+    perms: &'static [PermissionSet],
+}
+
+impl KnownFnTy {
+    /// Check that we annotated the right number of [`PermissionSet`]s
+    /// that corresponds to the number of raw pointers in [`Self::ty`].
+    #[cfg(test)]
+    const fn checked(self) -> Self {
+        let ty = self.ty.as_bytes();
+        let mut num_ptrs = 0;
+        let mut i = 0;
+        while i < ty.len() {
+            if ty[i] == b'*' {
+                num_ptrs += 1;
+            }
+            i += 1;
+        }
+        assert!(self.perms.len() == num_ptrs);
+        self
+    }
+
+    #[cfg(test)]
+    const fn named(self, name: &'static str) -> Self {
+        let Self { name: _, ty, perms } = self;
+        Self { name, ty, perms }
+    }
+}
+
+#[cfg(test)]
+macro_rules! known_fn_ty {
+    ($ty:ty: $perms:tt) => {{
+        KnownFnTy {
+            name: "",
+            ty: stringify!($ty),
+            perms: const_slice!(PermissionSet, perms_annotation!($perms)),
+        }
+        .checked()
+    }};
```

Comment:
Seems like this could take the name directly, instead of having a separate `named` method.  The tests below that use `known_fn_ty` without `named` could pass a dummy name instead.

Reply:
I tried that, but (and I forgot the exact reason why) it makes the macro ambiguous or the type inference not work or something.